Microsoft.Toolkit.HighPerformance/Helpers/Internals/SpanHelper.Count.cs

```diff
@@ -0,0 +1,305 @@
+
+        /// <summary>
+        /// Implements <see cref="Count{T}"/> with a vectorized search.
+        /// </summary>
+        [Pure]
+#if NETCOREAPP3_1
+        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
+#endif
+        private static unsafe int CountSimd<T>(ref T r0, IntPtr length, T value, IntPtr max)
+            where T : unmanaged, IEquatable<T>
+        {
+            int result = 0;
+
+            IntPtr offset = default;
+
+            // Skip the initialization overhead if there are not enough items
+            if ((byte*)length >= (byte*)Vector<T>.Count)
+            {
+                var vc = new Vector<T>(value);
+
+                do
+                {
+                    // Calculate the maximum sequential area that can be processed in
+                    // one pass without the risk of numeric overflow in the dot product
+                    // to sum the partial results. We also backup the current offset to
+                    // be able to track how many items have been processed, which lets
+                    // us avoid updating a third counter (length) in the loop body.
+                    IntPtr
+                        chunkLength = Min(length, max),
+                        initialOffset = offset;
+
+                    var partials = Vector<T>.Zero;
+
+                    while ((byte*)chunkLength >= (byte*)Vector<T>.Count)
+                    {
+                        ref T ri = ref Unsafe.Add(ref r0, offset);
+
+                        // Load the current Vector<T> register, and then use
+                        // Vector.Equals to check for matches. This API sets the
+                        // values corresponding to matching pairs to all 1s.
+                        // Since the input type is guaranteed to always be signed,
+                        // this means that a value with all 1s represents -1, as
+                        // signed numbers are represented in two's complement.
+                        // So we can just subtract this intermediate value to the
+                        // partial results, which effectively sums 1 for each match.
+                        var vi = Unsafe.As<T, Vector<T>>(ref ri);
+                        var ve = Vector.Equals(vi, vc);
+
+                        partials -= ve;
+
+                        chunkLength -= Vector<T>.Count;
+                        offset += Vector<T>.Count;
+                    }
+
+                    result += CastToInt(Vector.Dot(partials, Vector<T>.One));
+
+                    length = Subtract(length, Subtract(offset, initialOffset));
+                }
+                while ((byte*)length >= (byte*)Vector<T>.Count);
+            }
+
+            // Optional 8 unrolled iterations. This is only done when a single SIMD
+            // register can contain over 8 values of the current type, as otherwise
+            // there could never be enough items left after the vectorized path
+            if (Vector<T>.Count > 8 &&
```

Comment:
You could handle this via SIMD as well by backtracking, reading, and masking.

That is, lets say you are operating using `Vector128<uint>` and there are 3 elements remaining.
You can backtrack by 1 element, do a read, and then do your vectorized check as normal.
You can then mask out the additional 1, 2, or 3 elements that you read which shouldn't be included in the result.

This works for handling the first or last element and can also be used to align your data (whether it is pinned or you are optimistic the GC won't relocate it). It is generally faster than falling back to the length check + unrolled loop as well.

Reply:
That sounds super interesting, though I'm not exactly sure about how to properly do the backtrack/masking efficiently, do you have a reference (from CoreCLR or something else) that I can look up to as an example, at least as a starting point?

Also about this method, I saw that CoreCLR actually reverses the order and first alignes to `Vector<T>` boundaries, does the sequential step and then vectorizes, and eventually gets back to the sequential part for the remaining items ([here](https://github.com/dotnet/runtime/blob/d48be104ffbd44c1bd2607b3dea45050b9e3e214/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs#L205-L211)). Is that difference just a matter of style or is that recommended, and in that case should I do the same here? I'm curious because that CoreCLR code is still not pinning and not using the faster aligned vector load APIs not even for SSE/AVX, so I'm a bit confused. I gave that a try but after benchmarking the results were not really different (actually a bit worse in my case), is this expected? 🤔

Reply:
We aren't doing the trick in CoreCLR itself because there was concerns about it being confusing/unsafe since it reads outside the bounds of the memory. However, it's completely "safe" given how memory is laid out and access control works on the various pages.

You can see how it is used roughly here: https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.CpuMath/SseIntrinsics.cs#L36 and https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.CpuMath/SseIntrinsics.cs#L470

The premise is you just predefine 2, 4, 8, or 16 masks representing the elements you don't care about and then load the correct vector, selecting it via `maskArray + (remainder * Vector<T>.Count)`
You likewise move back by `Vector<T>.Count - remainder` when adjusting the first read.

Reply:
Ah, I see. Now I can understand why that's not used in CoreCLR, it does sound very confusing and potentially unsafe as you mentioned 🤣

Those links you shared are very interesting as well, thanks! I might give this a shot in a future PR if I can get a noticeable performance improvement in the profiler, for now I think I'll just leave this as it is if the logic itself is fine anyway, especially with the release deadline being so close. I'll definitely ping you again when I start messing with this part again though! 😊